enzyme/Enzyme/MLIR/Interfaces/Utils.cpp

```diff
@@ -0,0 +1,374 @@
+          }
+          if (*callee == "fscanf" || *callee == "__isoc99_fscanf") {
+            // Global read
+            effects.emplace_back(
+                MemoryEffects::Effect::get<MemoryEffects::Read>());
+
+            for (auto &&[idx, arg] :
+                 llvm::enumerate(cop.getArgOperandsMutable())) {
+              if (idx == 0) {
+                effects.emplace_back(MemoryEffects::Read::get(), &arg,
+                                     SideEffects::DefaultResource::get());
+                effects.emplace_back(MemoryEffects::Write::get(), &arg,
+                                     SideEffects::DefaultResource::get());
+              } else if (idx == 1) {
+                effects.emplace_back(MemoryEffects::Read::get(), &arg,
+                                     SideEffects::DefaultResource::get());
+              } else
+                effects.emplace_back(MemoryEffects::Write::get(), &arg,
+                                     SideEffects::DefaultResource::get());
+            }
+          }
+          if (*callee == "printf") {
+            // Global read
+            effects.emplace_back(
+                MemoryEffects::Effect::get<MemoryEffects::Write>());
+            for (auto &arg : cop.getArgOperandsMutable()) {
+              effects.emplace_back(MemoryEffects::Read::get(), &arg,
+                                   SideEffects::DefaultResource::get());
+            }
+          }
+          if (*callee == "free") {
+            for (auto &arg : cop.getArgOperandsMutable()) {
+              effects.emplace_back(MemoryEffects::Free::get(), &arg,
+                                   SideEffects::DefaultResource::get());
+            }
+          }
+          if (*callee == "strlen") {
+            for (auto &arg : cop.getArgOperandsMutable()) {
+              effects.emplace_back(MemoryEffects::Read::get(), &arg,
+                                   SideEffects::DefaultResource::get());
+            }
+          }
+        }
+      } else if (auto dop = dyn_cast<AutoDiffOp>(op)) {
+        // TODO: handle inter-procedural AutoDiffOp effects. Just skip for now.
+      } else if (auto dop = dyn_cast<ForwardDiffOp>(op)) {
+        // TODO: handle inter-procedural ForwardDiffOp effects. Just skip for
+        // now.
+      } else if (auto dop = dyn_cast<AutoDiffRegionOp>(op)) {
+        // TODO: handle intra-procedural AutoDiffRegionOp effects. Just skip for
+        // now.
+      } else {
+        // The operation does not have a memory effect interface, and we cannot
+        // obtain any result(consistent with the definition of
+        // `mlir::getEffectsRecursively`)
+        return std::nullopt;
+      }
+    }
+  }
+  return effects;
+}
+
+SmallVector<MemoryEffects::EffectInstance>
+collectFnEffects(FunctionOpInterface fnOp) {
+  SmallVector<MemoryEffects::EffectInstance> innerEffects;
+  for (auto &blk : fnOp.getBlocks()) {
+    for (auto &op : blk) {
+      auto opEffects = collectOpEffects(&op);
```

Comment:
This is wrong, if you define std::null_opt as not having effects, it needs to append all effect types here

Reply:
`std::nullopt` is that effects aren't defined at all, not having no effects(that's an empty vector)

Reply:
It's basically a re-implementation of https://github.com/llvm/llvm-project/blob/1127dd775426238cdc57b2581fbb91a0252e06ae/mlir/lib/Interfaces/SideEffectInterfaces.cpp#L346

Reply:
if effects aren't defined at all we need to be conservative and assume all effects